pkg/controller/ec2/vpcendpoint/setup.go

```diff
@@ -0,0 +1,176 @@
+			e.delete = c.delete
+			e.preCreate = preCreate
+			e.postCreate = postCreate
+			e.postObserve = postObserve
+			e.isUpToDate = isUpToDate
+			e.filterList = filterList
+		},
+	}
+	return ctrl.NewControllerManagedBy(mgr).
+		Named(name).
+		WithOptions(controller.Options{
+			RateLimiter: ratelimiter.NewDefaultManagedRateLimiter(rl),
+		}).
+		For(&svcapitypes.VPCEndpoint{}).
+		Complete(managed.NewReconciler(mgr,
+			resource.ManagedKind(svcapitypes.VPCEndpointGroupVersionKind),
+			managed.WithExternalConnecter(&connector{kube: mgr.GetClient(), opts: opts}),
+			managed.WithPollInterval(poll),
+			managed.WithLogger(l.WithValues("controller", name)),
+			managed.WithRecorder(event.NewAPIRecorder(mgr.GetEventRecorderFor(name)))))
+}
+
+type custom struct {
+	kube   client.Client
+	client svcsdkapi.EC2API
+}
+
+func preCreate(_ context.Context, cr *svcapitypes.VPCEndpoint, obj *svcsdk.CreateVpcEndpointInput) error {
+	// set external name as tag on the vpc endpoint
+	resType := "vpc-endpoint"
+	key := "Name"
+	value := meta.GetExternalName(cr)
+
+	spec := svcsdk.TagSpecification{
+		ResourceType: &resType,
+		Tags: []*svcsdk.Tag{
+			{
+				Key:   &key,
+				Value: &value,
+			},
+		},
+	}
+
+	obj.TagSpecifications = append(obj.TagSpecifications, &spec)
+	return nil
+}
+
+func postCreate(ctx context.Context, cr *svcapitypes.VPCEndpoint, obj *svcsdk.CreateVpcEndpointOutput, cre managed.ExternalCreation, err error) (managed.ExternalCreation, error) {
+	if err != nil {
+		return managed.ExternalCreation{}, err
+	}
+
+	// set vpc endpoint id as external name annotation on k8s object after creation
+	meta.SetExternalName(cr, aws.StringValue(obj.VpcEndpoint.VpcEndpointId))
+	cre.ExternalNameAssigned = true
+	return cre, nil
+}
+
+func postObserve(_ context.Context, cr *svcapitypes.VPCEndpoint, resp *svcsdk.DescribeVpcEndpointsOutput, obs managed.ExternalObservation, err error) (managed.ExternalObservation, error) {
+	if err != nil {
+		return managed.ExternalObservation{}, err
+	}
+
+	// Load DNS Entry as connection detail
+	if len(resp.VpcEndpoints[0].DnsEntries) != 0 && awsclients.StringValue(resp.VpcEndpoints[0].DnsEntries[0].DnsName) != "" {
+		obs.ConnectionDetails = managed.ConnectionDetails{
+			xpv1.ResourceCredentialsSecretEndpointKey: []byte(awsclients.StringValue(resp.VpcEndpoints[0].DnsEntries[0].DnsName)),
+		}
+	}
+
+	cr.Status.AtProvider.VPCEndpoint = generateVPCEndpointSDK(resp.VpcEndpoints[0])
+
+	switch awsclients.StringValue(resp.VpcEndpoints[0].State) {
+	case "available":
+		cr.SetConditions(xpv1.Available())
+	case "pending":
+		cr.SetConditions(xpv1.Creating())
+	case "pending-acceptance":
+		cr.SetConditions(xpv1.Creating())
+	case "deleted":
+		cr.SetConditions(xpv1.Unavailable())
+	case "deleting":
+		cr.SetConditions(xpv1.Deleting())
+	}
+
+	return obs, nil
+}
+
+func isUpToDate(*svcapitypes.VPCEndpoint, *svcsdk.DescribeVpcEndpointsOutput) (bool, error) {
+	return true, nil
```

Comment:
can you implement this? what fields are mutable?

also check pre/post update in another controller, like 
https://github.com/crossplane/provider-aws/blob/29466b1812a013378ba7bf845d712000c302408a/pkg/controller/docdb/dbsubnetgroup/setup.go#L137-L150

Reply:
This is now implemented.